tabbycat/notifications/views.py

```diff
@@ -37,13 +43,37 @@ def form_valid(self, form):
         return super().form_valid(form)
 
 
+class EmailEventWebhookView(TournamentMixin, View):
+
+    def post(self, request, *args, **kwargs):
+        if kwargs['key'] is not self.tournament.pref('email_hook_key'):
+            return HttpResponse(status_code=404) # 404: Not Found
+
+        data = json.loads(request.body)
+
+        for obj in data:
+            email = SentMessageRecord.objects.get(message_id=obj['smtp-id'])
+
+            dt = datetime.fromtimestamp(obj['timestamp'])
+            timestamp = timezone.make_aware(dt, timezone.utc)
+
+            EmailStatus.objects.create(email=email, timestamp=timestamp, event=obj['event'], data=obj)
```

Comment:
I don't know at what frequency the web-hook will fire, but if it fires each time an action occurs, there's no benefit in using `.bulk_create()`.

For the email `.get`, more in #848, but it would be more efficient if `email__message_id=` could be used. (Can it?)

Reply:
It's every 30 seconds or when it hits 768 kB, whichever happens first: https://sendgrid.com/docs/API_Reference/Event_Webhook/getting_started_event_webhook.html#-Data-Storage
On a quick back-of-the-envelope calculation, 768 kB is about 2,000 events, so it's probably jsut once every 30 seconds or so.

I believe field lookups aren't allowed in `create()` calls, and even if they were, I think it'd involve some intelligence at the Python level that still translates to multiple database hits?

Reply:
But `id` fields still work? I'm getting more tempted to replace the primary key now. (Also they are generated by Django, not the mail server)

Reply:
Oh yeah, if `author` is a foreign key in `Book` then `Book.objects.create(author_id=...)` is a valid move. Moving the rest of this to #868 lol

Reply:
What does iron-manning have to do with this?

Reply:
Whoops, I meant #848, sorry.

You avoid an N+1 inefficiency like this:

``` python
records = SentMessageRecord.objects.filter(message_id__in=[obj['smtp-id'] for obj in data])
record_lookup = {smr.message_id: smr.id for smr in records}
statuses = []

for obj in data:
    dt = datetime.fromtimestamp(obj['timestamp'])
    timestamp = timezone.make_aware(dt, timezone.utc)
    email = record_lookup.get(obj['smtp-id'], None)
    if email is None:
        continue
    statuses.append(EmailStatus(email=email, timestamp=timestamp, event=obj['event'], data=obj))

EmailStatus.objects.bulk_create(statuses)
```

That'll be exactly 2 queries, every time.

Reply:
Checked the Django docs. New solution is in the `EmailStatus` foreign key field, use `to_field=message_id` and `db_constraint=False`.

Reply:
It's worth saying that you'll need to do that whether you use a natural primary key (the message ID) or a surrogate primary key (the numeric one), for the reasons I discuss in #848.